csharp/src/Ice/ConnectionFactory.cs

```diff
@@ -142,6 +150,33 @@ internal OutgoingConnectionFactory(Communicator communicator)
                 }
             }
 
+            lock (_mutex)
+            {
+                if (_disposeTask != null)
+                {
+                    throw new CommunicatorDisposedException();
+                }
+
+                // Purge expired hint failures
+                DateTime expirationDate = DateTime.Now - TimeSpan.FromSeconds(5);
+                var expiredFailures = _connectFailures.Where(entry => entry.Value <= expirationDate).Select(
+                    entry => entry.Key).ToList();
+                foreach (IConnector expired in expiredFailures)
+                {
+                    _connectFailures.Remove(expired);
+                }
+
+                // Exclude connectors that where already tried, order the remaining connectors moving connectors with
+                // recent failures to the end of the list.
+                connectors = connectors.Where(item => !excludedConnectors.Contains(item.Connector)).OrderBy(
+                    item => _connectFailures.TryGetValue(item.Connector, out DateTime value) ? value : default).ToList();
+
+                if (connectors.Count == 0)
+                {
+                    throw new NoMoreReplicasException();
+                }
```

Comment:
Could we just use the NoEndpointException here instead of adding a new exception?

Reply:
We can, the only difference is that NoEndpointException can be throw before the first invocation, in which case we should rethrow it to user code.

Reply:
Removed `NoMoreReplicasException` and use `NoEndpointException` instead